src/module/EMDoublePairProduction.cpp

```diff
@@ -69,8 +75,19 @@ void EMDoublePairProduction::performInteraction(Candidate *candidate) const {
 	Random &random = Random::instance();
 	Vector3d pos = random.randomInterpolatedPosition(candidate->previous.getPosition(), candidate->current.getPosition());
 
-	candidate->addSecondary( 11, Ee, pos);
-	candidate->addSecondary(-11, Ee, pos);
+	double f = Ee / E;
+	double w0 = candidate->getWeight();
+
+	if (haveElectrons) {
```

Comment:
Does the difference in the two if conditions -- (1-f)^\alpha compared to f^alpha -- introduce a difference between particles and anti-particles or am I missing something?

When the energy fraction f is small then electrons are produced more often than positrons, aren't they?

Reply:
It does. It increases efficiency, but since we are not tracking all the particles charge is not conserved.

Reply:
Would it be possible to just dice the sign of the particle ID?
In this way the efficiency should stay the same but on average for constant energy fractions f the number of positrons and electrons should converge to the same value.

Reply:
Take a look at my comment below. There is no bias towards either charge, since the ones that are kept are chosen randomly based solely on the fraction of the total energy they take. The more energetic ones are kept, and their weights increased, whereas the low-energy ones will rarely be added to the array of secondaries.
If electrons and positrons have equal probability of being produced, then two random numbers are drawn and, on average, the simulation will be charge-neutral.
In other words, the higher the thinning, the more the most energetic particles are valued, since there are only a few of them compared to the many that are produced at lower energies.

Reply:
I agree with you that is is true for the the other module but here I am not yet convinced.

Let's assume we have a maximal thinning `thinning=1` and the primary gamma-ray has an energy `E=4*m_e c^2 ` -- which might not be realistic due to production thresholds but makes my argument more clear.

In this case, the energy for positron and electron is `Ee=(E - 2*m_e c^2)/2=m_e c^2 ` and the energy fraction is `f=1/4`. 

Then we would have a 75%=(1-f) chance of producing an electron and a 25% chance for a positron. When I read the code correctly, these chances are constant for a given input energy E. 

This is different from the other module, where Ee!=Ep and Ee will be different in each call of the performInteraction function.

One last thought:
In reality it's possible that this treatment is not problematic at all.
Assuming that `E>>m_e c^2` --> f=1/2 which will lead to a equal production rate for positrons and electrons.

Reply:
This is not exactly true.
This implementation follows Lee 1996 mentioned in the comments. He says:
"Therefore, I use the assumption where one pair of the two carries all the initial energy and two particles in the pair share the energy equally."
Therefore, f=1/2 (not 1/4) for one electron and positron, and 0 for the others.

This is weird, I agree. DPP in this case is just the usual PP (it only produces one pair, effectively), but the cross sections are different, hence it's importance. There might be better implementations.

I could have just set f=0.5 from the beginning. I kept it this way because in the future we might want to improve this implementation and the it will require less work.

And this treatment is okay exactly for the reason you said: f=1/2 at high energies, so regardless of the thinning we get + and - charges balanced.

Reply:
@rafaelab Thanks for you detailed explanations. I think I just misunderstood the implementation regardless of the thinning or not. I should have taken a look into the original paper (Lee 1996). I'll try to come up with a short comment that we might add to the code, to make the procedure more clear. But this should not block your PR.

Reply:
But now that you mention, although they are the same in this case, one line should be Ee, and the other Ep, for consistency. I will change that.